lxd/db/generate/README.md

```diff
@@ -0,0 +1,187 @@
+# `lxd-generate`
+
+## Introduction
+
+`lxd-generate` is a database statement and associated `go` function generator
+for LXD and related projects. `lxd-generate` utilizes `go`'s code generation
+directives (`//go:generate ...`) alongside go's [ast](https://pkg.go.dev/go/ast)
+package for parsing the syntax tree for go structs and variables. We use
+`lxd-generate` for the majority of our SQL statements and database interactions
+on the `go` side for consistency and predictability.
+
+## Usage
+
+### Initialization
+
+Generally the first thing we will want to do for any newly generated file is to
+establish the command, the target file, and ensure the file has been cleared of
+content:
+
+```go
+//go:generate -command mapper lxd-generate db mapper -t instances.mapper.go
+//go:generate mapper reset -i -b "//go:build linux && cgo && !agent"
+
+```
+
+This will initiate a call to `lxd-generate db mapper -t instances.mapper.go` on
+each call to the go generation directive `mapper`.
+
+### SQL Statement generation
+
+SQL generation supports the following SQL statement types:
+Type                                  | Description
+:---                                  | :----
+`objects`                             | Creates a basic SELECT statement of the form `SELECT <columns> FROM <table> ORDER BY <columns>`.
+`objects-by-<FIELD>-and-<FIELD>...`   | Parses a pre-existing SELECT statement of the form produced by`objects`, and appends a `WHERE` clause with the given fields located in the associated struct.
+`create`                              | Creates a basic INSERT statement of the form `INSERT INTO <table> VALUES`.
+`create-or-replace`                   | Creates a basic INSERT statement of the form `INSERT OR REPLACE INTO <table> VALUES`.
+`delete-by-<FIELD>-and-<FIELD>...`    | Creates a DELETE statement of the form `DELETE FROM <table> WHERE <constraint>` where the constraint is based on the given fields of the associated struct.
+`id`                                  | Creates a basic SELECT statement that returns just the internal ID of the table.
+`rename`                              | Creates an UPDATE statement that updates the primary key of a table: `UPDATE <table> SET <primary key> WHERE <primary key = ?>`.
+`update`                              | Creates an UPDATE statement of the form `UPDATE <table> SET <all columns> WHERE <primary key = ?>`.
+
+#### Examples
+
+```go
+//go:generate mapper stmt -e instance objects
+//go:generate mapper stmt -e instance objects-by-Name-and-Project
+//go:generate mapper -e instance create
+//go:generate mapper -e instance update
+//go:generate mapper -e instance delete-by-Name-and-Project
+```
+
+`//go:generate mapper stmt -e instance objects table=table_name`
```

Comment:
Does this need a sub-section called Arguments or similar as it feels a bit "bolted on" at the end?

Reply:
And does this apply to all of the SQL statement generation commands above or just a subset of them?

Reply:
Sure, there are some other generation directive config keys that I've given their own section now. They reference the `EntityTable`, `ReferenceTable`, `MapTable`, and `AssociationTable` definitions below.

Also yes, the `table=<name>` config key applies to all `stmt` commands.